crates/m3/src/gadgets/hash/keccak/lookedup.rs

```diff
@@ -0,0 +1,759 @@
+			+ PackedExtension<B32>,
+		PackedSubfield<P, B8>: PackedTransformationFactory<PackedSubfield<P, B8>>,
+		PackedSubfield<P, B32>: PackedFieldIndexable<Scalar = B32>,
+	{
+		for (k, trace) in permutation_traces.iter().enumerate() {
+			// Gather all batch round traces for the batch number.
+			let brt = PerBatchLens::new(trace, self.batch_no);
+
+			// Fill in round_const witness with the corresponding round constants
+			let mut round_const: std::cell::RefMut<'_, [u64]> =
+				index.get_mut_as(self.round_const)?;
+			for track in 0..TRACKS_PER_BATCH {
+				round_const[TRACKS_PER_BATCH * k + track] = brt[track].rc;
+			}
+			drop(round_const);
+
+			for x in 0..5 {
+				let mut c: std::cell::RefMut<'_, [u64]> = index.get_mut_as(self.c[x])?;
+				let mut c_shift: std::cell::RefMut<'_, [u64]> =
+					index.get_mut_as(self.c_shift[x])?;
+				let mut d: std::cell::RefMut<'_, [u64]> = index.get_mut_as(self.d[x])?;
+				for track in 0..TRACKS_PER_BATCH {
+					let cell_pos = TRACKS_PER_BATCH * k + track;
+					c[cell_pos] = brt[track].c[x];
+					c_shift[cell_pos] = c[cell_pos].rotate_left(1);
+					d[cell_pos] = brt[track].d[x];
+				}
+
+				for y in 0..5 {
+					let mut state_in: std::cell::RefMut<'_, [u64]> =
+						index.get_mut_as(self.state_in[(x, y)])?;
+					let mut state_out: std::cell::RefMut<'_, [u64]> =
+						index.get_mut_as(self.state_out[(x, y)])?;
+					let mut a_theta: std::cell::RefMut<'_, [u64]> =
+						index.get_mut_as(self.a_theta[(x, y)])?;
+
+					// b[0,0] is defined as a_theta[0,0].
+					//
+					// That means two things:
+					// 1. The value is assigned by `a_theta`.
+					// 2. We have to skip mutably borrowing it here because that would overlap with
+					//    the mutable borrow of `a_theta` above.
+					let mut b: Option<std::cell::RefMut<'_, [u64]>> = if (x, y) != (0, 0) {
+						Some(index.get_mut_as(self.b[(x, y)])?)
+					} else {
+						None
+					};
+					for track in 0..TRACKS_PER_BATCH {
+						let cell_pos = TRACKS_PER_BATCH * k + track;
+						state_in[cell_pos] = brt[track].state_in[(x, y)];
+						state_out[cell_pos] = brt[track].state_out[(x, y)];
+						a_theta[cell_pos] = brt[track].a_theta[(x, y)];
+						if let Some(ref mut b) = b {
+							b[cell_pos] = brt[track].b[(x, y)];
+						}
+					}
+				}
+			}
+		}
+		for xy in 0..25 {
+			let x = xy % 5;
+			let y = xy / 5;
+			let mut merged: std::cell::RefMut<'_, [B32]> =
+				index.get_scalars_mut(self.merged[(x, y)])?;
+
+			let b1: std::cell::Ref<'_, [B8]> = index.get_as(self.b[(x + 1, y)])?;
+
+			let b2: std::cell::Ref<'_, [B8]> = index.get_as(self.b[(x + 2, y)])?;
+			let mut b1_and_b2: std::cell::RefMut<'_, [B8]> =
+				index.get_mut_as(self.b1_and_b2[(x, y)])?;
+			for i in 0..b2.len() {
+				// B[x,y] xor ((not B[x+1,y]) and B[x+2,y])
+				let in_a = b1[i].val();
+				let in_b = b2[i].val();
+				let output = in_a & in_b;
+				b1_and_b2[i] = output.into();
+				merged[i] = merge_bitand_vals(in_a, in_b, output).into();
+			}
+		}
```

Comment:
Nit, small one: literally everywhere in this module we use `for x in 0..5 { for y in 0..5 { } }` and only here it's the inverse.